lib/async_hooks.js

```diff
@@ -197,6 +200,137 @@ class AsyncResource {
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        contexts[i]._propagate(resource, triggerResource);
+      }
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        delete resource[contexts[i].kResourceStore];
+      }
+      idResourcePairing.delete(asyncId);
+    }
+  }
+});
+
+class AsyncContext {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  enterSync() {
+    asyncContextHook.enable();
+    if (!contexts.includes(this)) {
+      contexts.push(this);
+    }
+    let resource = getCurrentResource();
+    if (!resource) {
+      // there is no resource, we need to create a fake one
+      resource = {};
+      idResourcePairing.set(executionAsyncId(), new WeakReference(resource));
+      process.nextTick(() => { resource }); // this prevents the fake resource from being garbage collected too soon
+    }
+    const store = new Map();
+    resource[this.kResourceStore] = store;
+    return store;
+  }
+
+  exitSync() {
+    const resource = getCurrentResource();
+    if (resource) {
+      delete resource[this.kResourceStore];
+    }
+  }
+
+  runAndReturn(callback, ...args) {
+    const store = this.enterSync();
+    try {
+      return callback(store, ...args);
+    } finally {
+      this.exitSync();
+    }
+  }
+
+  exitAndReturn(callback, ...args) {
+    const resource = getCurrentResource();
+    if (!resource) {
+      return callback(...args); // let's do nothing here, there is no context
+    }
+    const store = resource[this.kResourceStore];
+    delete resource[this.kResourceStore];
+    try {
+      return callback(...args);
+    } finally {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const resource = getCurrentResource();
+    if (resource) {
+      return resource[this.kResourceStore];
+    }
+  }
+
+  /**
+   * after this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  run(callback, ...args) {
+    const store = this.enterSync();
+    process.nextTick(() => callback(store, ...args));
```

Comment:
Shouldn't the `enterSync()` be inside the `process.nextTick(...)` callback so it doesn't leak the entered context into following sync code after calling `run(...)`?

Reply:
Good catch, it's either that, either an `exitSync`missing after the `process.nextTick`. I will fix that.

Reply:
fixed! (and assertion added to test it)